File: python/phasemodel/tests/test_hmc_biased.py
"""
Test script for hybrid montecarlo sampling.
"""

# make sure phasemodel package is in path
import sys,os
cwd = os.path.abspath(os.path.dirname(__file__))
sys.path.insert(0,os.path.join(cwd,"..",".."))

import numpy as np
import phasemodel
import phasemodel.utils as utils 
import phasemodel.hmc2 as hmc2  
import phasemodel.f_energy as en

def test_hmc_biased():
    # generate random coupling
    dim = 5
    
    K_true = np.random.randn(dim+1,dim+1)+1j*np.random.randn(dim+1,dim+1)
    K_true[np.diag(np.ones(dim+1,bool))] = 0
    K_true = .5*(K_true+np.conj(K_true.T))

    # convert coupling from complex 3x3 to real 6x6 matrix
    M = phasemodel.kappa2m(K_true);

    # some settings
    opts = hmc2.opt(
        nsamples = 10**3,
        nomit = 10**3,
        steps = 50,
        stepadj = .15,
        persistence = False)

    # generate test data
    utils.tic()
    samps = hmc2.hmc2(en.f_phasedist_biased,np.zeros(dim),opts,en.g_phasedist_biased,M)
    data = utils.smod(samps.T)
    utils.toc()

    # fit test data
    K_fit = phasemodel.fit_model_biased(data);

    print K_true
    print K_fit

    K_error = (abs(K_true-K_fit)).mean()

    print """

    mean-absolute-difference = %6.8f; expect: 0.2

    """%(K_error)


if __name__ == '__main__':
    test_hmc_biased()

File: matlab/test_hmc_biased.m
% Test script for hybrid montecarlo sampling
% 
% In order to use hybrid montecarlo sampling,
% the functions hmc2.m and hmc2_opt.m are needed.
% Both files are distributed separately here:
% http://www.lce.hut.fi/research/mm/mcmcstuff/mcmcstuff.tar.gz

%% setup

addpath('f_energy');

%% load data

% load('testdata/three_phases.mat','K_true','data');
% sz = size(K_true,1);

%%

dim = 3;
K_true = randn(dim+1,dim+1)+1j*randn(dim+1,dim+1);
K_true(logical(eye(dim+1))) = 0;
K_true = .5*(K_true+conj(transpose(K_true)));
sz = size(K_true,1)-1;
% K_true(2:end,2:end) = 0;

%% convert coupling from complex 3x3 to real 6x6 matrix

M = kappa2m(K_true);

%% some settings

nsamples = 10^3;
burnin = 10^3;
lf_steps = 50;
step_sz = .15;
persistence = 0;

%% sample some data

opts = hmc2_opt();
opts.nsamples = nsamples;
opts.nomit = burnin;
opts.steps = lf_steps;
opts.stepadj = step_sz;
opts.persistence = persistence;

[samps, E, diagn] = hmc2('f_phasedist_biased',zeros(sz,1),opts,'g_phasedist_biased',M);
[nsamples,dim] = size(samps);
data = zeros(dim+1,nsamples);
data(2:end,1:end) = smod(transpose(samps));

%% fit data with double precision

K_fit = fit_model(data);

%% plot results

hval = max(max(abs(K_fit(:))),max(abs(K_true(:))));

figure(1)
subplot(121)
imagesc(abs(K_true),[-1 1]*hval)
title('True coupling (K\_true)')
axis square off
subplot(122)
imagesc(abs(K_fit),[-1 1]*hval)
title('Estimated coupling, matlab (K\_fit)')
axis square off

K_true
K_fit

K_error = mean(abs(K_true(:)-K_fit(:)));
fprintf('\n double precision');
fprintf('\n mean-absolute-difference = %6.8f; expect: ~ 0.07\n',K_error);

